src/index/addrindex.cpp

```diff
@@ -0,0 +1,225 @@
+{}
+
+BaseIndex::DB& AddrIndex::GetDB() const { return *m_db; }
+
+bool AddrIndex::DB::ReadAddrIndex(const uint64_t addr_id,
+                                  std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_found,
+                                  const bool filter_by_value,
+                                  const uint64_t value_wanted){
+    bool found_tx = false; // return true only if at least one transaction was found
+    const std::pair<char, uint64_t> key_prefix = std::make_pair(DB_ADDRINDEX, addr_id);
+    std::unique_ptr<CDBIterator> iter(NewIterator());
+
+    iter->Seek(key_prefix);
+    while (iter->Valid()) {
+        std::pair<std::pair<char, uint64_t>, CDiskTxPos> key;
+        uint64_t value;
+        if (!iter->GetKey(key) || !iter->GetValue(value) || key.first != key_prefix) break;
+
+        if  (!filter_by_value || (filter_by_value && value == value_wanted)) {
+            found_tx = true;
+            keys_found.emplace_back(key);
+        }
+
+        iter->Next();
+    }
+
+    return found_tx;
+}
+
+void AddrIndex::DB::RemoveKeys(const std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_to_remove) {
+    CDBBatch batch(*this);
+    for (const auto& key : keys_to_remove) {
+        batch.Erase(key);
+    }
+    WriteBatch(batch);
+}
+
+AddrIndex::AddrIndex(size_t n_cache_size, bool f_memory, bool f_wipe)
+    : m_db(MakeUnique<AddrIndex::DB>(n_cache_size, f_memory, f_wipe)){}
+
+AddrIndex::~AddrIndex() {}
+
+uint64_t AddrIndex::GetAddrID(const CScript& script) {
+    uint256 hashed_script;
+
+    CSHA256 hasher;
+    hasher.Write((unsigned char*)&(*script.begin()), script.end() - script.begin());
+    hasher.Finalize(hashed_script.begin());
+
+    return hashed_script.GetUint64(0);
+}
+
+bool AddrIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<uint64_t, CDiskTxPos>> positions;
+    positions.reserve(2 * block.vtx.size());  // Most transactions have at least 1 input and 1 output.
+
+    // Index addresses of spent outputs if txindex is enabled,
+    for (const auto& tx : block.vtx) {
+        for (const auto tx_out : tx->vout){
+            positions.emplace_back(GetAddrID(tx_out.scriptPubKey), pos);
+        }
+
+        if (g_txindex && !tx->IsCoinBase()) {
+            for (const auto tx_in : tx->vin) {
+                CTransactionRef tx;
+                uint256 block_hash;
+
+                if (!g_txindex->FindTx(tx_in.prevout.hash, block_hash, tx)) {
+                    // Both addrindex and txindex may be syncing in parallel, and addrindex might
+                    // be ahead of txindex. We let txindex sync first so that addrindex can continue
+                    // after it.
+                    while (!g_txindex->IsInSyncWithMainChain()) {
+                        MilliSleep(1000); //TODO: find a less arbitrary sleep time.
```

Comment:
In commit "Introduce address index" (3c7cc3c705b08828dc8dc919e53343df78568ebd)

This sleep + BlockUntilSyncedToCurrentChain call is pretty hacky but probably ok in principle. I think @jimpo could easily tell you how to improve it. I would move this code out of the `AddrIndex` class and into a `g_txindex->WaitForSync()` or similar method, and probably replace the sleep with a condition variable wait.

Or maybe this code will be unnecessary with #14035 undo data?